File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> mapSet = new HashMap<>();
        int left = 0;
        int maxLength = 0;

        for(int right = 0; right< s.length(); right++){
            if(mapSet.containsKey(s.charAt(right))){
                left = Math.max(left, mapSet.get(s.charAt(right))+1);
            }
            mapSet.put(s.charAt(right), right);
            maxLength = Math.max(maxLength, right -left + 1);
        }
        return maxLength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
       left = maxLength = 0;
       charSet = set()

       for right in range(len(s)):
            while s[right] in charSet:
                charSet.remove(s[left])
                left += 1

            charSet.add(s[right])
            maxLength = max(maxLength, right - left + 1)

       return maxLength

        